Zend/zend_ast.c

```diff
@@ -449,12 +449,16 @@ ZEND_API int zend_ast_evaluate(zval *result, zend_ast *ast, zend_class_entry *sc
 			} else {
 				zval tmp;
 
-				zend_fetch_dimension_const(&tmp, &op1, &op2, (ast->attr == ZEND_DIM_IS) ? BP_VAR_IS : BP_VAR_R);
+				if (!Z_ISERROR(op1)) {
+					zend_fetch_dimension_const(&tmp, &op1, &op2, (ast->attr == ZEND_DIM_IS) ? BP_VAR_IS : BP_VAR_R);
 
-				if (UNEXPECTED(Z_ISREF(tmp))) {
-					ZVAL_DUP(result, Z_REFVAL(tmp));
+					if (UNEXPECTED(Z_ISREF(tmp))) {
+						ZVAL_DUP(result, Z_REFVAL(tmp));
+					} else {
+						ZVAL_DUP(result, &tmp);
+					}
 				} else {
-					ZVAL_DUP(result, &tmp);
+					ZVAL_DUP(result, &op1);
```

Comment:
There is a problem here, because this code propagates the ERROR, even though not all using code will be able to handle it. I saw that you accounted for it in zend_update_constant, but this result may also be only one subexpression in a larger AST evaluation. For example, if you have `const FOO = BAR[0] + 1`, where `BAR[0]` yields ERROR, the evaluator for the binary operation will not be able to deal with the ERROR operand.

I think the most elegant way to fix this is to add a flag to the AST_DIM node, which determines whether an ERROR should be propagated or not, similar to the existing ZEND_DIM_IS flag. The right place to set this flag would be in https://github.com/php/php-src/blob/master/Zend/zend_compile.c#L8513. For example, you could modify the code handling DIM_IS to something like this:

```
if (ast->child[0]->kind == ZEND_AST_DIM) {
    if (ast->attr == ZEND_DIM_IS) {
        /* Propagate DIM_IS flag */
        ast->child[0]->attr = ZEND_DIM_IS;
    } else {
        ast->child[0]->attr = ZEND_DIM_PROPAGATE_ERROR;
    }
}
```

And then only propagate the error if the flag is set, otherwise replace it with null.